# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [2]:
import pandas as pd
import numpy as np

In [3]:
horses = pd.read_csv('horse_data.csv',header=None, na_values='?')
horses_columns_1 = pd.read_csv('horse_data.names.csv',header=None)
horses.columns = list(map(lambda x: x.strip(),list(horses_columns_1[1]) ))

In [4]:
horses.describe()

surgery?         Age  Hospital Number  rectal temperature  \
count  299.000000  300.000000     3.000000e+02          240.000000   
mean     1.397993    1.640000     1.085889e+06           38.167917   
std      0.490305    2.173972     1.529801e+06            0.732289   
min      1.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289040e+05           37.800000   
50%      1.000000    1.000000     5.303055e+05           38.200000   
75%      2.000000    1.000000     5.347275e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  respiratory rate  temperature of extremities  \
count  276.000000        242.000000                  244.000000   
mean    71.913043         30.417355                    2.348361   
std     28.630557         17.642231                    1.045054   
min     30.000000          8.000000                    1.000000   
25%     48.000000         18.500000                    1.000000   
50%     64.000000         24.500000                    3.000000   
75%     88.000000         36.000000                    3.000000   
max    184.000000         96.000000                    4.000000   

       peripheral pulse  mucous membranes  capillary refill time  ...  \
count        231.000000        253.000000             268.000000  ...   
mean           2.017316          2.853755               1.305970  ...   
std            1.042428          1.620294               0.477629  ...   
min            1.000000          1.000000               1.000000  ...   
25%            1.000000          1.000000               1.000000  ...   
50%            2.000000          3.000000               1.000000  ...   
75%            3.000000          4.000000               2.000000  ...   
max            4.000000          6.000000               3.000000  ...   

       packed cell volume  total protein  abdominocentesis appearance  \
count          271.000000     267.000000                   135.000000   
mean            46.295203      24.456929                     2.037037   
std             10.419335      27.475009                     0.804905   
min             23.000000       3.300000                     1.000000   
25%             38.000000       6.500000                     1.000000   
50%             45.000000       7.500000                     2.000000   
75%             52.000000      57.000000                     3.000000   
max             75.000000      89.000000                     3.000000   

          outcome  abdomcentesis total protein  surgical lesion?  \
count  102.000000                   299.000000        300.000000   
mean     3.019608                     1.551839          1.363333   
std      1.968567                     0.737187          0.481763   
min      0.100000                     1.000000          1.000000   
25%      2.000000                     1.000000          1.000000   
50%      2.250000                     1.000000          1.000000   
75%      3.900000                     2.000000          2.000000   
max     10.100000                     3.000000          2.000000   

       type of lesion_1  type of lesion_2  type of lesion_3     cp_data  
count        300.000000        300.000000        300.000000  300.000000  
mean        3657.880000         90.226667          7.363333    1.670000  
std         5399.513513        649.569234        127.536674    0.470998  
min            0.000000          0.000000          0.000000    1.000000  
25%         2111.750000          0.000000          0.000000    1.000000  
50%         2673.500000          0.000000          0.000000    2.000000  
75%         3209.000000          0.000000          0.000000    2.000000  
max        41110.000000       7111.000000       2209.000000    2.000000  

[8 rows x 28 columns]

## Описание Столбцов
1.  surgery - проводилась ли операция 1 - Да, 2 - Нет {1-2} [качественная]
1.  Age - 1 - взрослая, 0 - молодая лошадь {1-2} [качественная]
1.  Hospital Number  - цифровой код, который может быть не уникальным [качественная]
1.  rectal temperature - [количественная] 
1.  pulse - {1-40}[количественная]
1.  respiratory rate  - [количественная] 
1.  temperature of extremities - Температура конечностей {1-4} [качественная]
1.  peripheral pulse - периферический пульс {1-4} [качественная]
1.  mucous membranes - слизистая оболочка {1-6} [качественная]
1.  capillary refill time - время наполнения капилляров {1-2} [качественная]
1.  pain - ощущает ли лошадь боль {1-5} [качественная]
1.  peristalsis - перистатика {1-4} [качественная]
1.  abdominal distension - вздутие живота {1-4} [качественная] (сразу показывает на перистатику. 3, 4 означает 4 перистатики) 
1.  nasogastric tube - назогастальный зонд - {1-3} [качественная]
1.  nasogastric reflux - назогастальная обратный ток(отрыжка) {1-3} [качественная]
1.  nasogastric reflux PH - назогастальная обратный ток(отрыжка) PH среды {1-14} [количественная]
1.  rectal examination - испражнения {1-4} [качественная]
1.  abdomen - брюхо {1-4} [качественная] 4-5 указывает на хирургическое вмешательство
1.  packed cell volume - объем осжденых эритроцитов {30-50} [количественная] >50 указывает на обезвоживание
1.  total protein - содержание белка {6-7.5} [количественная] >7.5 указывает на обезвоживание
1.  abdominocentesis appearance - пункция живота {1-3} [качественная]
1.  abdominocentesis total protein - содержание белка  [количественная] повышен содержание указывает на заболевание
1.  outcome - Результат {1-3} [качественная]
1.  surgical lesions - хирургические осложнения {1-2} [качественная]
1.  type of lesion_1 - тип осложнения {00,1-11} [качественная]
1.  type of lesion_2 - тип осложнения {1-4} [качественная]
1.  type of lesion_3 - тип осложнения {0-2} [качественная]
1.  cp_data - информация о предыдущих заболеваниях {1-2} [качественная]


для метрики необходимо различать качественные или количественные значения. Количественные данные это:  
   * 4, 5, 6, 16, 19, 20, 22  

Качественные:   
   * все остальное.

Для количественных:   
   * mode, mediana, outline, quadratic dispersion  (standart deviation), quantile, outliers
  
  
Для качественных:   
   * mode

In [6]:
for i in [4, 5, 6, 16, 19, 20, 22]:
    column = horses.columns[i].strip()
    print(column)
    mode = horses[horses.columns[i]].mode()[0]
    print(f'чаще всего встречается у "{column}" - {mode}')
    mean = horses[horses.columns[i]].mean()
    print(f'Среднее значение у "{column}" - {mean:.2f}')  
    median = horses[horses.columns[i]].median()
    print(f'Медиана у " {column}" - {median}') 
    standart_deviation = horses[horses.columns[i]].std()
    quadratic_dispersion = horses[horses.columns[i]].std()*2
    print(f'Стандартное отклонение у "{column}" - {standart_deviation:.2f}') 
    print(f'Квадрат стандартного отклонения у "{column}" - {quadratic_dispersion:.2f}')  
    se = standart_deviation / (300**0.5) # стандартное отклонение средней = средняя / корень кол-ва измерений (корень 300)
    confidential_interval_95 = median + (1.96 * se)
    confidential_interval_99 = median + (2.58 * se)
    print(f'95% Доверительный интервал у "{column}" - {confidential_interval_95:.2f}') 
    print(f'99% Доверительный интервал у "{column}" - {confidential_interval_99:.2f}') 
    #Найти квантиль, Размах, выбросы
    q1 = horses[horses.columns[i]].quantile(0.25)
    q2 = horses[horses.columns[i]].quantile(0.5)
    q3 = horses[horses.columns[i]].quantile(0.75)
    iqr = q3 - q1 
    lower_outlier = q1 - 1.5 * iqr
    upper_outlier = q3 + 1.5 * iqr
    A={} #Создаем словарик из всех значений столбца, чтобы потом итерируясь по нему найти выбросы
    for j in horses.index:
        A[j] = horses[horses.columns[i]][j]
    #Находим выбросы
    outliers = {j:A[j] for j in A.keys() if (A[j] > upper_outlier) or (A[j] < lower_outlier)}
    #horses.iloc[list(outliers.keys()),:] # Если нужно вывести датафрейм с выбросами
    print(f'Верхняя граница уса "{column}" - {upper_outlier:.2f}')
    print(f'Нижняя граница уса "{column}" - {lower_outlier:.2f}')  
    print(f'Выбросы у "{column}" - {outliers}')
    print()
    #break

for i in range (1,28,1):
    if i in [4, 5, 6, 16, 19, 20, 22]: continue
    column = horses.columns[i].strip()
    print(column)
    mode = horses[horses.columns[i]].mode()[0]
    print(f'чаще всего встречается у "{column}" - {mode}')
    print()
    #break

pulse
чаще всего встречается у "pulse" - 48.0
Среднее значение у "pulse" - 71.91
Медиана у " pulse" - 64.0
Стандартное отклонение у "pulse" - 28.63
Квадрат стандартного отклонения у "pulse" - 57.26
95% Доверительный интервал у "pulse" - 67.24
99% Доверительный интервал у "pulse" - 68.26
Верхняя граница уса "pulse" - 148.00
Нижняя граница уса "pulse" - -12.00
Выбросы у "pulse" - {3: 164.0, 41: 150.0, 55: 160.0, 255: 184.0, 275: 150.0}

respiratory rate
чаще всего встречается у "respiratory rate" - 20.0
Среднее значение у "respiratory rate" - 30.42
Медиана у " respiratory rate" - 24.5
Стандартное отклонение у "respiratory rate" - 17.64
Квадрат стандартного отклонения у "respiratory rate" - 35.28
95% Доверительный интервал у "respiratory rate" - 26.50
99% Доверительный интервал у "respiratory rate" - 27.13
Верхняя граница уса "respiratory rate" - 62.25
Нижняя граница уса "respiratory rate" - -7.75
Выбросы у "respiratory rate" - {3: 84.0, 39: 96.0, 41: 72.0, 82: 80.0, 84: 80.0, 103: 68.0, 

In [188]:
%%time #Расчет занимает 20ms
upper_outliner = 100
lower_outliner = 50

remove_outliners = horses[horses[horses.columns[4]].between(lower_outliner, upper_outliner)]
outliners = pd.concat([horses,remove_outliners]).drop_duplicates(keep = False)
outliners

Wall time: 20 ms


surgery?    Age    Hospital Number    rectal temperature    pulse  \
2           2.0      1             530334                  38.3     40.0   
3           1.0      9            5290409                  39.1    164.0   
4           2.0      1             530255                  37.3    104.0   
5           2.0      1             528355                   NaN      NaN   
6           1.0      1             526802                  37.9     48.0   
..          ...    ...                ...                   ...      ...   
288         1.0      1             529428                   NaN      NaN   
290         2.0      1             535054                  38.6     45.0   
294         2.0      1             533902                  38.5     40.0   
295         1.0      1             533886                   NaN    120.0   
299         1.0      1             534618                  37.2     40.0   

       respiratory rate    temperature of extremities    peripheral pulse  \
2                  24.0                           1.0                 1.0   
3                  84.0                           4.0                 1.0   
4                  35.0                           NaN                 NaN   
5                   NaN                           2.0                 1.0   
6                  16.0                           1.0                 1.0   
..                  ...                           ...                 ...   
288                 NaN                           NaN                 NaN   
290                16.0                           2.0                 1.0   
294                16.0                           1.0                 1.0   
295                70.0                           4.0                 NaN   
299                20.0                           NaN                 NaN   

       mucous membranes   capillary refill time  ...   packed cell volume  \
2                   3.0                     1.0  ...                 33.0   
3                   6.0                     2.0  ...                 48.0   
4                   6.0                     2.0  ...                 74.0   
5                   3.0                     1.0  ...                  NaN   
6                   1.0                     1.0  ...                 37.0   
..                  ...                     ...  ...                  ...   
288                 NaN                     NaN  ...                  NaN   
290                 2.0                     1.0  ...                 43.0   
294                 1.0                     1.0  ...                 37.0   
295                 4.0                     2.0  ...                 55.0   
299                 NaN                     NaN  ...                 36.0   

      total protein   abdominocentesis appearance   outcome  \
2               6.7                           NaN       NaN   
3               7.2                           3.0       5.3   
4               7.4                           NaN       NaN   
5               NaN                           NaN       NaN   
6               7.0                           NaN       NaN   
..              ...                           ...       ...   
288             NaN                           NaN       NaN   
290            58.0                           NaN       NaN   
294            67.0                           NaN       NaN   
295            65.0                           NaN       NaN   
299            62.0                           1.0       1.0   

      abdomcentesis total protein   surgical lesion?   type of lesion_1  \
2                             1.0                  2                  0   
3                             2.0                  1               2208   
4                             2.0                  2               4300   
5                             1.0                  2                  0   
6                             1.0                  1               3124   
..                            ...    

In [187]:
%%time #Расчет занимает 0ns
upper_outlier = 100
lower_outlier = 50

A={} 
for i in horses.index:
    A[i] = horses[horses.columns[4]][i]
outliers = {j:A[j] for j in A.keys() if (A[j] >= upper_outlier) or (A[j] <= lower_outlier)}

horses.iloc[list(outliers.keys()),:] 

Wall time: 0 ns


surgery?    Age    Hospital Number    rectal temperature    pulse  \
2           2.0      1             530334                  38.3     40.0   
3           1.0      9            5290409                  39.1    164.0   
4           2.0      1             530255                  37.3    104.0   
6           1.0      1             526802                  37.9     48.0   
12          1.0      1             528031                  37.2     42.0   
..          ...    ...                ...                   ...      ...   
290         2.0      1             535054                  38.6     45.0   
294         2.0      1             533902                  38.5     40.0   
295         1.0      1             533886                   NaN    120.0   
298         1.0      1             530612                  36.5    100.0   
299         1.0      1             534618                  37.2     40.0   

       respiratory rate    temperature of extremities    peripheral pulse  \
2                  24.0                           1.0                 1.0   
3                  84.0                           4.0                 1.0   
4                  35.0                           NaN                 NaN   
6                  16.0                           1.0                 1.0   
12                 12.0                           2.0                 1.0   
..                  ...                           ...                 ...   
290                16.0                           2.0                 1.0   
294                16.0                           1.0                 1.0   
295                70.0                           4.0                 NaN   
298                24.0                           3.0                 3.0   
299                20.0                           NaN                 NaN   

       mucous membranes   capillary refill time  ...   packed cell volume  \
2                   3.0                     1.0  ...                 33.0   
3                   6.0                     2.0  ...                 48.0   
4                   6.0                     2.0  ...                 74.0   
6                   1.0                     1.0  ...                 37.0   
12                  1.0                     1.0  ...                  NaN   
..                  ...                     ...  ...                  ...   
290                 2.0                     1.0  ...                 43.0   
294                 1.0                     1.0  ...                 37.0   
295                 4.0                     2.0  ...                 55.0   
298                 3.0                     1.0  ...                 50.0   
299                 NaN                     NaN  ...                 36.0   

      total protein   abdominocentesis appearance   outcome  \
2               6.7                           NaN       NaN   
3               7.2                           3.0       5.3   
4               7.4                           NaN       NaN   
6               7.0                           NaN       NaN   
12              7.0                           NaN       NaN   
..              ...                           ...       ...   
290            58.0                           NaN       NaN   
294            67.0                           NaN       NaN   
295            65.0                           NaN       NaN   
298             6.0                           3.0       3.4   
299            62.0                           1.0       1.0   

      abdomcentesis total protein   surgical lesion?   type of lesion_1  \
2                             1.0                  2                  0   
3                             2.0                  1               2208   
4                             2.0                  2               4300   
6                             1.0                  1               3124   
12                            1.0                  2               4124   
..                            ...    

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [18]:
outliners_dict = {}
for i in [4, 5, 6, 16, 19, 20, 22]:
    print(horses.columns[i])
    median = horses[horses.columns[i]].median()
    print(f'Медиана у " {horses.columns[i]}" - {median}') 
    standart_deviation = horses[horses.columns[i]].std()
    quadratic_dispersion = horses[horses.columns[i]].std()*2
    print(f'Стандартное отклонение у "{horses.columns[i]}" - {standart_deviation:.2f}') 
    print(f'Квадрат стандартного отклонения у "{horses.columns[i]}" - {quadratic_dispersion:.2f}')  
    se = standart_deviation / (300**0.5) # стандартное отклонение средней = средняя / корень кол-ва измерений (корень 300)
    confidential_interval_95 = median + (1.96 * se)
    confidential_interval_99 = median + (2.58 * se)
    print(f'95% Доверительный интервал у "{horses.columns[i]}" - {confidential_interval_95:.2f}') 
    print(f'99% Доверительный интервал у "{horses.columns[i]}" - {confidential_interval_99:.2f}')
    
    q1 = horses[horses.columns[i]].quantile(0.25)
    q2 = horses[horses.columns[i]].quantile(0.5)
    q3 = horses[horses.columns[i]].quantile(0.75)
    iqr = q3 - q1 
    lower_outlier = q1 - 1.5 * iqr
    upper_outlier = q3 + 1.5 * iqr
    A={} #Создаем словарик из всех значений столбца, чтобы потом итерируясь по нему найти выбросы
    for j in horses.index:
        A[j] = horses[horses.columns[i]][j]
    #Находим выбросы
    outliers = {j:A[j] for j in A.keys() if (A[j] > upper_outlier) or (A[j] < lower_outlier)}
    print(f'Верхняя граница уса "{horses.columns[i]}" - {upper_outlier:.2f}')
    print(f'Нижняя граница уса "{horses.columns[i]}" - {lower_outlier:.2f}')  
    print(f'Выбросы у "{horses.columns[i]}" - {outliers}')
    print()
    outliners_dict[horses.columns[i]] = list(outliers.keys())
    #break

pulse
Медиана у " pulse" - 64.0
Стандартное отклонение у "pulse" - 28.27
Квадрат стандартного отклонения у "pulse" - 56.55
95% Доверительный интервал у "pulse" - 67.20
99% Доверительный интервал у "pulse" - 68.21
Верхняя граница уса "pulse" - 148.00
Нижняя граница уса "pulse" - -12.00
Выбросы у "pulse" - {55: 160.0, 255: 184.0, 275: 150.0}

respiratory rate
Медиана у " respiratory rate" - 24.5
Стандартное отклонение у "respiratory rate" - 17.64
Квадрат стандартного отклонения у "respiratory rate" - 35.28
95% Доверительный интервал у "respiratory rate" - 26.50
99% Доверительный интервал у "respiratory rate" - 27.13
Верхняя граница уса "respiratory rate" - 62.25
Нижняя граница уса "respiratory rate" - -7.75
Выбросы у "respiratory rate" - {3: 84.0, 39: 96.0, 41: 72.0, 82: 80.0, 84: 80.0, 103: 68.0, 106: 96.0, 120: 66.0, 125: 68.0, 186: 90.0, 208: 80.0, 229: 70.0, 244: 88.0, 255: 84.0, 265: 68.0, 269: 90.0, 295: 70.0}

temperature of extremities
Медиана у " temperature of extremities" - 3.

### pulse
При первом взгляде складывается ощущение, что выбросы за гранью верхнего уса и 99% доверительного интервала. Но если присмотреться к группам лошадей, разбитым по возрасту и переживанию боли, пульс сильно разнится:

In [8]:
horses.groupby(['Age','pain']).pulse.median()

Age  pain
1    1.0      48.0
     2.0      72.0
     3.0      55.0
     4.0      77.5
     5.0      88.0
9    1.0      92.0
     2.0     124.0
     3.0     120.0
     4.0     128.0
     5.0     136.0
Name: pulse, dtype: float64

Посчитаем верхний ус:

In [9]:
def calc_box_wiskers(data):
    q1 = data.pulse.quantile(0.25)
    q3 = data.pulse.quantile(0.75)
    iqr = q3 - q1 
    upper_outliner = q3 + 1.5 * iqr
    return upper_outliner

In [10]:
horses.groupby(['Age','pain']).apply(calc_box_wiskers)

Age  pain
1    1.0      61.375
     2.0     161.500
     3.0      94.250
     4.0     151.250
     5.0     141.000
9    1.0      92.000
     2.0     139.000
     3.0     200.000
     4.0     219.000
     5.0     188.000
dtype: float64

Выборка, которая попала в выброс: 

In [11]:
horses.iloc[outliners_dict['pulse'] ,:][['Age', 'pain','pulse','surgery?']]

Age  pain  pulse  surgery?
3      9   2.0  164.0       1.0
41     9   NaN  150.0       2.0
55     9   3.0  160.0       1.0
255    9   4.0  184.0       1.0
275    9   5.0  150.0       1.0

У всех выбросов, кроме одного, пульс находится в границе верхнего уса, что говорит о правдоподобности таких значений. В описании сказано, что если болезненное расстройство, то пульс может быть повышен, тем более, что 4 из 5 лошадям из этой выборки проводили хирургическое вмешательство.  
У лошади № 41 пульс 150, не зная категорию боли, но предполагая, что раз операции не было, стало быть боль не была высокой категории максимум категории 2, а  пульс 150 - выше чем для боли 2 категории (граница верхнего уса -139) - стало быть данное значение выброс и его лучше заменить на значение границы верхнего уса.   То же самое со значением пульса лошади №3, категория боли 2, а значение пульса 164 выше границы верхнего уса - стало быть данное значение выброс и его лучше заменить на значение границы верхнего уса.

In [12]:
horses.iat[3,4] = 139
horses.iat[41,4] = 139

In [13]:
horses.iat[41,4]

139.0

### лошади с учащенным дыханием
В описании указано, что польза от данной метрики ['respiratory rate'] сомнительна изза высокой волатильности значений, поэтому оставить все значения не повлияет на какие-либо выводы.

In [19]:
horses.iloc[outliners_dict['respiratory rate'] ,:] 

surgery?  Age  Hospital Number  rectal temperature  pulse  \
3         1.0    9          5290409                39.1  139.0   
39        1.0    9          5277409                39.2  146.0   
41        2.0    9          5288249                39.0  139.0   
82        1.0    9          5290759                38.1  100.0   
84        1.0    1           529849                37.8   60.0   
103       1.0    9          5292489                38.0  140.0   
106       1.0    1           528668                38.3   52.0   
120       1.0    1           528682                39.4   54.0   
125       1.0    1           528461                38.0   42.0   
186       1.0    1          5278331                39.3   64.0   
208       1.0    1           535338                37.8   88.0   
229       1.0    9           534597                38.5  120.0   
244       1.0    9          5291409                38.2  124.0   
255       1.0    9          5294539                38.8  184.0   
265       2.0    1           528590                 NaN   56.0   
269       1.0    1           527883                39.3   64.0   
295       1.0    1           533886                 NaN  120.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
3                84.0                         4.0               1.0   
39               96.0                         NaN               NaN   
41               72.0                         NaN               NaN   
82               80.0                         3.0               1.0   
84               80.0                         1.0               3.0   
103              68.0                         1.0               1.0   
106              96.0                         NaN               3.0   
120              66.0                         1.0               1.0   
125              68.0                         4.0               1.0   
186              90.0                         2.0               3.0   
208              80.0                         3.0               3.0   
229              70.0                         NaN               NaN   
244              88.0                         1.0               3.0   
255              84.0                         1.0               NaN   
265              68.0                         3.0               1.0   
269              90.0                         2.0               3.0   
295              70.0                         4.0               NaN   

     mucous membranes  capillary refill time  ...  packed cell volume  \
3                 6.0                    NaN  ...                48.0   
39                NaN                    NaN  ...                 NaN   
41                NaN                    NaN  ...                47.0   
82                2.0                    NaN  ...                36.0   
84                2.0                    NaN  ...                40.0   
103               1.0                    NaN  ...                39.0   
106               3.0                    NaN  ...                43.0   
120               2.0                    NaN  ...                39.0   
125               1.0                    NaN  ...                41.0   
186               1.0                    NaN  ...                39.0   
208               5.0                    NaN  ...                64.0   
229               NaN                    NaN  ...                35.0   
244               2.0                    NaN  ...                47.0   
255               1.0                    NaN  ...                33.0   
265               1.0                    NaN  ...                40.0   
269               1.0                    NaN  ...                39.0   
295               4.0                    NaN  ...                55.0   

     total protein  abdominocentesis appearance  outcome  \
3              7.2                          3.0      NaN   
39             NaN                          NaN      NaN   
41             8.5                        

### Дополнение
Не входит в объем задачи, но в категорийных качественных метриках, есть значения, которые не соответствуют описанию, например в значении метрики outcome всего три значения 1, 2, 3. В датафрейме значения 0.1 и 5.3, которые не подходят под описание.   
Самое правильное их будет удалить.

In [15]:
S = {
    'surgery?' : {1,2},
    'temperature of extremities' : {1,2,3,4},
    'peripheral pulse' : {1,2,3,4},
    'mucous membranes' : {1,2,3,4,5,6},
    'capillary refill time' : {1-2},
    'pain' : {1,2,3,4,5},
    'peristalsis' : {1,2,3,4},
    'abdominal distension' : {1,2,3,4},
    'nasogastric tube' : {1,2,3},
    'nasogastric reflux' : {1,2,3},
    'rectal examination' : {1,2,3,4},
    'abdomen' : {1,2,3,4} ,
    'abdominocentesis appearance' : {1,2,3,4},
    'outcome' : {1,2,3},
    'surgical lesions' : {1, 2},
    'cp_data' : {1, 2}}

In [16]:
for name in horses.columns:
    if name in S: 
        
        for i in range(0,300,1):
            if horses.at[i,name] not in S[name]:
                horses.at[i,name] = 'NaN'
        
horses       

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0         2.0    1           530101                38.5   66.0   
1         1.0    1           534817                39.2   88.0   
2         2.0    1           530334                38.3   40.0   
3         1.0    9          5290409                39.1  139.0   
4         2.0    1           530255                37.3  104.0   
..        ...  ...              ...                 ...    ...   
295       1.0    1           533886                 NaN  120.0   
296       2.0    1           527702                37.2   72.0   
297       1.0    1           529386                37.5   72.0   
298       1.0    1           530612                36.5  100.0   
299       1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
0                28.0                         3.0               3.0   
1                20.0                         NaN               NaN   
2                24.0                         1.0               1.0   
3                84.0                         4.0               1.0   
4                35.0                         NaN               NaN   
..                ...                         ...               ...   
295              70.0                         4.0               NaN   
296              24.0                         3.0               2.0   
297              30.0                         4.0               3.0   
298              24.0                         3.0               3.0   
299              20.0                         NaN               NaN   

     mucous membranes  capillary refill time  ...  packed cell volume  \
0                 NaN                    NaN  ...                45.0   
1                 4.0                    NaN  ...                50.0   
2                 3.0                    NaN  ...                33.0   
3                 6.0                    NaN  ...                48.0   
4                 6.0                    NaN  ...                74.0   
..                ...                    ...  ...                 ...   
295               4.0                    NaN  ...                55.0   
296               4.0                    NaN  ...                44.0   
297               4.0                    NaN  ...                60.0   
298               3.0                    NaN  ...                50.0   
299               NaN                    NaN  ...                36.0   

     total protein  abdominocentesis appearance  outcome  \
0              8.4                          NaN      NaN   
1             85.0                          2.0      2.0   
2              6.7                          NaN      NaN   
3              7.2                          3.0      NaN   
4              7.4                          NaN      NaN   
..             ...                          ...      ...   
295           65.0                          NaN      NaN   
296            NaN                          3.0      NaN   
297            6.8                          NaN      NaN   
298            6.0                          3.0      NaN   
299           62.0                          1.0      1.0   

     abdomcentesis total protein  surgical lesion?  type of lesion_1  \
0                            2.0                 2             11300   
1                            3.0                 2              2208   
2                            1.0                 2                 0   
3                            2.0                 1              2208   
4                            2.0                 2              4300   
..                           ...               ...               ...   
295                          3.0                 2              3205   
296                          3.0                 1              2208   
297                          2.0                 1              3205   
298                          1.0         

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [20]:
for colum in horses.columns:
    aver_missing = horses[colum].isna().mean()
    print(f'{colum} - {aver_missing:.1%}')

surgery? - 0.3%
Age - 0.0%
Hospital Number - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
respiratory rate - 19.3%
temperature of extremities - 18.7%
peripheral pulse - 23.0%
mucous membranes - 15.7%
capillary refill time - 100.0%
pain - 18.3%
peristalsis - 14.7%
abdominal distension - 18.7%
nasogastric tube - 34.7%
nasogastric reflux - 35.3%
nasogastric reflux PH - 82.3%
rectal examination - feces - 34.0%
abdomen - 65.7%
packed cell volume - 9.7%
total protein - 11.0%
abdominocentesis appearance - 55.0%
outcome - 85.0%
abdomcentesis total protein - 0.3%
surgical lesion? - 0.0%
type of lesion_1 - 0.0%
type of lesion_2 - 0.0%
type of lesion_3 - 0.0%
cp_data - 0.0%


### Индивидуально каждый столбец просматривать:

### surgery?

In [21]:
horses[horses['surgery?'].isna()][['surgery?','surgical lesion?']] 

surgery?  surgical lesion?
132       NaN                 1

очевидно что surgery было, потому что были осложнения при surgery

In [22]:
horses.at[132,'surgery?'] = 1

### rectal temperature

In [24]:
horses[horses['rectal temperature'].isna()].head()

surgery?  Age  Hospital Number  rectal temperature  pulse  \
5        2.0    1           528355                 NaN    NaN   
7        1.0    1           529607                 NaN   60.0   
8        2.0    1           530051                 NaN   80.0   
16       1.0    9          5301219                 NaN  128.0   
28       1.0    1          5279442                 NaN    NaN   

    respiratory rate  temperature of extremities  peripheral pulse  \
5                NaN                         2.0               1.0   
7                NaN                         3.0               NaN   
8               36.0                         3.0               4.0   
16              36.0                         3.0               3.0   
28               NaN                         NaN               NaN   

    mucous membranes  capillary refill time  ...  packed cell volume  \
5                3.0                    NaN  ...                 NaN   
7                NaN                    NaN  ...                44.0   
8                3.0                    NaN  ...                38.0   
16               4.0                    NaN  ...                53.0   
28               NaN                    NaN  ...                 NaN   

    total protein  abdominocentesis appearance  outcome  \
5             NaN                          NaN      NaN   
7             8.3                          NaN      NaN   
8             6.2                          NaN      NaN   
16            7.8                          3.0      NaN   
28            NaN                          NaN      NaN   

    abdomcentesis total protein  surgical lesion?  type of lesion_1  \
5                           1.0                 2                 0   
7                           2.0                 1              2208   
8                           3.0                 1              3205   
16                          2.0                 2              1400   
28                          2.0                 2              4300   

    type of lesion_2  type of lesion_3  cp_data  
5                  0                 0        2  
7                  0                 0        2  
8                  0                 0        2  
16                 0                 0        1  
28                 0                 0        2  

[5 rows x 28 columns]

In [25]:
horses.groupby(['Age','pain'])['rectal temperature'].median()

Age  pain
1    1.0     38.2
     2.0     38.3
     3.0     38.1
     4.0     38.0
     5.0     38.0
9    1.0     38.0
     2.0     38.8
     3.0     38.4
     4.0     38.8
     5.0     38.3
Name: rectal temperature, dtype: float64

Заполню средней тепературой в зависимтости от возраста и категории боли

In [26]:
horses['rectal temperature'].fillna( horses.groupby(['Age','pain'])['rectal temperature'].transform('median'), inplace = True  )
horses[horses['rectal temperature'].isna()] #Почему он поменял не все значения не знаю. Синтаксис как на лекции.

surgery?  Age  Hospital Number  rectal temperature  pulse  \
7         1.0    1           529607                 NaN   60.0   
28        1.0    1          5279442                 NaN    NaN   
56        1.0    1           528872                 NaN    NaN   
68        2.0    1           529777                 NaN   52.0   
74        1.0    9          5292929                 NaN    NaN   
87        2.0    1           528006                 NaN   40.0   
96        2.0    1           530319                 NaN   84.0   
152       1.0    1           533836                 NaN  100.0   
175       1.0    1           527929                 NaN    NaN   
207       1.0    1           530297                 NaN    NaN   
216       2.0    1           528977                 NaN    NaN   
227       1.0    9          5289419                 NaN    NaN   
274       1.0    1           534624                 NaN   76.0   
288       1.0    1           529428                 NaN    NaN   
293       1.0    1           534004                 NaN   78.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
7                 NaN                         3.0               NaN   
28                NaN                         NaN               NaN   
56                NaN                         NaN               NaN   
68               16.0                         1.0               NaN   
74                NaN                         NaN               NaN   
87               16.0                         NaN               NaN   
96               36.0                         NaN               NaN   
152               NaN                         3.0               3.0   
175               NaN                         NaN               NaN   
207               NaN                         NaN               NaN   
216               NaN                         NaN               NaN   
227               NaN                         NaN               NaN   
274               NaN                         NaN               3.0   
288               NaN                         NaN               NaN   
293              24.0                         3.0               3.0   

     mucous membranes  capillary refill time  ...  packed cell volume  \
7                 NaN                    NaN  ...                44.0   
28                NaN                    NaN  ...                 NaN   
56                NaN                    NaN  ...                24.0   
68                3.0                    NaN  ...                55.0   
74                NaN                    NaN  ...                37.0   
87                NaN                    NaN  ...                50.0   
96                3.0                    NaN  ...                44.0   
152               4.0                    NaN  ...                68.0   
175               NaN                    NaN  ...                 NaN   
207               NaN                    NaN  ...                45.0   
216               NaN                    NaN  ...                 NaN   
227               NaN                    NaN  ...                45.0   
274               NaN                    NaN  ...                 NaN   
288               NaN                    NaN  ...                 NaN   
293               3.0                    NaN  ...                43.0   

     total protein  abdominocentesis appearance  outcome  \
7              8.3                          NaN      NaN   
28             NaN                          NaN      NaN   
56             6.7                          NaN      NaN   
68             7.2                          NaN      NaN   
74             4.9                          NaN      NaN   
87             7.0                          2.0      NaN   
96             8.5                          NaN      NaN   
152           64.0                          3.0      2.0   
175            NaN                          NaN      NaN   
207            8.7                        

### pulse

In [30]:
horses[horses['pulse'].isna()]

surgery?  Age  Hospital Number  rectal temperature  pulse  \
5         2.0    1           528355                38.3    NaN   
28        1.0    1          5279442                 NaN    NaN   
52        2.0    1           529483                38.2    NaN   
56        1.0    1           528872                 NaN    NaN   
58        1.0    1           528298                38.0    NaN   
74        1.0    9          5292929                 NaN    NaN   
78        1.0    1           530693                38.0    NaN   
83        1.0    1          5279822                38.0    NaN   
93        2.0    1           530310                38.0    NaN   
115       2.0    1           533723                38.1    NaN   
117       2.0    1          5290482                39.5    NaN   
126       1.0    1           530384                38.7    NaN   
136       1.0    1           530693                38.0    NaN   
151       1.0    1           529399                39.3    NaN   
159       2.0    1           528134                38.3    NaN   
160       1.0    1           527916                38.0    NaN   
174       2.0    1           518476                38.0    NaN   
175       1.0    1           527929                 NaN    NaN   
182       1.0    1          5279822                38.0    NaN   
204       1.0    1           529528                39.2    NaN   
207       1.0    1           530297                 NaN    NaN   
216       2.0    1           528977                 NaN    NaN   
227       1.0    9          5289419                 NaN    NaN   
288       1.0    1           529428                 NaN    NaN   

     respiratory rate  temperature of extremities  peripheral pulse  \
5                 NaN                         2.0               1.0   
28                NaN                         NaN               NaN   
52                NaN                         1.0               1.0   
56                NaN                         NaN               NaN   
58               20.0                         4.0               3.0   
74                NaN                         NaN               NaN   
78                NaN                         3.0               3.0   
83               24.0                         3.0               3.0   
93                NaN                         3.0               3.0   
115              40.0                         3.0               1.0   
117               NaN                         3.0               3.0   
126               NaN                         3.0               1.0   
136               NaN                         3.0               3.0   
151               NaN                         4.0               3.0   
159              12.0                         1.0               1.0   
160               NaN                         NaN               NaN   
174               NaN                         3.0               4.0   
175               NaN                         NaN               NaN   
182              24.0                         3.0               3.0   
204              23.0                         3.0               1.0   
207               NaN                         NaN               NaN   
216               NaN                         NaN               NaN   
227               NaN                         NaN               NaN   
288               NaN                         NaN               NaN   

     mucous membranes  capillary refill time  ...  packed cell volume  \
5                 3.0                    NaN  ...                 NaN   
28                NaN                    NaN  ...                 NaN   
52                3.0                    NaN  ...                43.0   
56                NaN                    NaN  ...                24.0   
58                3.0                    NaN  ...                53.0   
74                NaN                    NaN  ...                37.0   
78                3.0                    NaN  ...                46

In [31]:
horses.groupby(['surgery?','pain'])['pulse'].median()

surgery?  pain
1.0       1.0     48.0
          2.0     84.0
          3.0     64.0
          4.0     73.5
          5.0     88.0
2.0       1.0     48.0
          2.0     60.0
          3.0     54.0
          4.0     84.0
          5.0     96.0
Name: pulse, dtype: float64

In [35]:
horses['pulse'].fillna( horses.groupby(['surgery?','pain'])['pulse'].transform('median'), inplace = True  )
horses[horses['pulse'].isna()] #Почему он поменял не все значения не знаю. Синтаксис как на лекции.

surgery?  Age  Hospital Number  rectal temperature  pulse  \
28        1.0    1          5279442                 NaN    NaN   
56        1.0    1           528872                 NaN    NaN   
74        1.0    9          5292929                 NaN    NaN   
175       1.0    1           527929                 NaN    NaN   
207       1.0    1           530297                 NaN    NaN   
216       2.0    1           528977                 NaN    NaN   
227       1.0    9          5289419                 NaN    NaN   
288       1.0    1           529428                 NaN    NaN   

     respiratory rate  temperature of extremities  peripheral pulse  \
28                NaN                         NaN               NaN   
56                NaN                         NaN               NaN   
74                NaN                         NaN               NaN   
175               NaN                         NaN               NaN   
207               NaN                         NaN               NaN   
216               NaN                         NaN               NaN   
227               NaN                         NaN               NaN   
288               NaN                         NaN               NaN   

     mucous membranes  capillary refill time  ...  packed cell volume  \
28                NaN                    NaN  ...                 NaN   
56                NaN                    NaN  ...                24.0   
74                NaN                    NaN  ...                37.0   
175               NaN                    NaN  ...                 NaN   
207               NaN                    NaN  ...                45.0   
216               NaN                    NaN  ...                 NaN   
227               NaN                    NaN  ...                45.0   
288               NaN                    NaN  ...                 NaN   

     total protein  abdominocentesis appearance  outcome  \
28             NaN                          NaN      NaN   
56             6.7                          NaN      NaN   
74             4.9                          NaN      NaN   
175            NaN                          NaN      NaN   
207            8.7                          NaN      NaN   
216            NaN                          NaN      NaN   
227            6.5                          2.0      NaN   
288            NaN                          NaN      NaN   

     abdomcentesis total protein  surgical lesion?  type of lesion_1  \
28                           2.0                 2              4300   
56                           1.0                 1              3112   
74                           2.0                 1             11124   
175                          1.0                 1              2208   
207                          2.0                 1              2208   
216                          1.0                 2                 0   
227                          1.0                 1              9000   
288                          1.0                 1              2124   

     type of lesion_2  type of lesion_3  cp_data  
28                  0                 0        2  
56                  0                 0        2  
74                  0                 0        2  
175                 0                 0        2  
207                 0                 0        2  
216                 0                 0        2  
227                 0                 0        1  
288                 0                 0        2  

[8 rows x 28 columns]

### respiratory rate - 19.3%   
*usefulness is doubtful due to the great fluctuations*

In [39]:
horses['respiratory rate'].median()

24.5

In [38]:
horses['respiratory rate'].fillna( horses['respiratory rate'].median(), inplace = True  )
horses[horses['respiratory rate'].isna()] 

Empty DataFrame
Columns: [surgery?, Age, Hospital Number, rectal temperature, pulse, respiratory rate, temperature of extremities, peripheral pulse, mucous membranes, capillary refill time, pain, peristalsis, abdominal distension, nasogastric tube, nasogastric reflux, nasogastric reflux PH, rectal examination - feces, abdomen, packed cell volume, total protein, abdominocentesis appearance, outcome, abdomcentesis total protein, surgical lesion?, type of lesion_1, type of lesion_2, type of lesion_3, cp_data]
Index: []

[0 rows x 28 columns]

### temperature of extremities - 18.7%


In [56]:
horses[horses['temperature of extremities'].isna()]

surgery?  Age  Hospital Number  rectal temperature  pulse  \
1         1.0    1           534817                39.2   88.0   
4         2.0    1           530255                37.3  104.0   
17        2.0    1           526639                37.5   48.0   
23        1.0    9           534998                38.3  130.0   
25        2.0    1           529518                37.8   60.0   
28        1.0    1          5279442                 NaN    NaN   
36        2.0    1           529493                38.3  112.0   
37        1.0    1           533847                37.8   72.0   
39        1.0    9          5277409                39.2  146.0   
41        2.0    9          5288249                39.0  139.0   
53        2.0    1           530544                38.6   40.0   
56        1.0    1           528872                 NaN    NaN   
72        1.0    1          5299253                37.7   56.0   
74        1.0    9          5292929                 NaN    NaN   
81        1.0    1           534756                37.3   40.0   
87        2.0    1           528006                 NaN   40.0   
90        2.0    1           534197                38.0   52.0   
95        1.0    1           535158                37.3   48.0   
96        2.0    1           530319                 NaN   84.0   
106       1.0    1           528668                38.3   52.0   
113       1.0    1           527933                36.8   60.0   
121       1.0    1           534556                38.3   80.0   
128       1.0    1           535330                37.4   84.0   
129       1.0    1           527698                38.4   49.0   
130       1.0    1           528964                37.8   30.0   
141       2.0    1           522979                36.0   42.0   
148       1.0    1           529272                38.3  132.0   
149       2.0    1           535314                38.7   48.0   
160       1.0    1           527916                38.0   73.5   
161       1.0    1           530431                38.3   42.0   
165       1.0    1           534925                39.4  120.0   
169       1.0    1           535085                38.0   65.0   
175       1.0    1           527929                 NaN    NaN   
177       1.0    1           523190                38.0   68.0   
181       2.0    1           528305                37.8   48.0   
190       1.0    1           528742                37.1   40.0   
202       1.0    1           534073                37.5   48.0   
207       1.0    1           530297                 NaN    NaN   
210       1.0    1           530401                37.9   68.0   
216       2.0    1           528977                 NaN    NaN   
217       1.0    1          5279441                38.0   76.0   
227       1.0    9          5289419                 NaN    NaN   
229       1.0    9           534597                38.5  120.0   
232       1.0    1          5262543                38.5   30.0   
239       1.0    1           534183                38.1   70.0   
260       2.0    1           534933                38.0   40.0   
262       2.0    9          5305629                38.6   88.0   
263       1.0    1           528743                37.1   75.0   
268       1.0    1           534403                38.2   42.0   
274       1.0    1           534624                 NaN   76.0   
280       1.0    1           534644                38.6   48.0   
283       1.0    1           527758                38.2   72.0   
287       2.0    9          5305129                39.5   84.0   
288       1.0    1           529428                 NaN    NaN   
289       1.0    1           529126                38.0   50.0   
299       1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
1                20.0                         NaN               NaN   
4                35.0                         NaN               NaN   
17               24.0        

In [55]:
horses.groupby(['surgery?','pain'])['temperature of extremities'].median()

surgery?  pain
1.0       1.0     1.0
          2.0     3.0
          3.0     2.0
          4.0     3.0
          5.0     3.0
2.0       1.0     2.0
          2.0     3.0
          3.0     2.0
          4.0     3.0
          5.0     3.0
Name: temperature of extremities, dtype: float64

In [57]:
horses['temperature of extremities'].fillna( horses.groupby(['surgery?','pain'])['temperature of extremities'].transform('median'), inplace = True  )
horses[horses['temperature of extremities'].isna()]

surgery?  Age  Hospital Number  rectal temperature  pulse  \
4         2.0    1           530255                37.3  104.0   
17        2.0    1           526639                37.5   48.0   
25        2.0    1           529518                37.8   60.0   
28        1.0    1          5279442                 NaN    NaN   
36        2.0    1           529493                38.3  112.0   
39        1.0    9          5277409                39.2  146.0   
41        2.0    9          5288249                39.0  139.0   
53        2.0    1           530544                38.6   40.0   
56        1.0    1           528872                 NaN    NaN   
72        1.0    1          5299253                37.7   56.0   
74        1.0    9          5292929                 NaN    NaN   
87        2.0    1           528006                 NaN   40.0   
96        2.0    1           530319                 NaN   84.0   
106       1.0    1           528668                38.3   52.0   
113       1.0    1           527933                36.8   60.0   
129       1.0    1           527698                38.4   49.0   
130       1.0    1           528964                37.8   30.0   
141       2.0    1           522979                36.0   42.0   
161       1.0    1           530431                38.3   42.0   
165       1.0    1           534925                39.4  120.0   
175       1.0    1           527929                 NaN    NaN   
202       1.0    1           534073                37.5   48.0   
207       1.0    1           530297                 NaN    NaN   
216       2.0    1           528977                 NaN    NaN   
217       1.0    1          5279441                38.0   76.0   
227       1.0    9          5289419                 NaN    NaN   
229       1.0    9           534597                38.5  120.0   
232       1.0    1          5262543                38.5   30.0   
262       2.0    9          5305629                38.6   88.0   
268       1.0    1           534403                38.2   42.0   
274       1.0    1           534624                 NaN   76.0   
283       1.0    1           527758                38.2   72.0   
287       2.0    9          5305129                39.5   84.0   
288       1.0    1           529428                 NaN    NaN   
299       1.0    1           534618                37.2   40.0   

     respiratory rate  temperature of extremities  peripheral pulse  \
4                35.0                         NaN               NaN   
17               24.0                         NaN               NaN   
25               42.0                         NaN               NaN   
28               24.5                         NaN               NaN   
36               16.0                         NaN               3.0   
39               96.0                         NaN               NaN   
41               72.0                         NaN               NaN   
53               20.0                         NaN               NaN   
56               24.5                         NaN               NaN   
72               24.5                         NaN               NaN   
74               24.5                         NaN               NaN   
87               16.0                         NaN               NaN   
96               36.0                         NaN               NaN   
106              96.0                         NaN               3.0   
113              28.0                         NaN               NaN   
129              24.5                         NaN               NaN   
130              12.0                         NaN               NaN   
141              30.0                         NaN               NaN   
161              24.0                         NaN               NaN   
165              48.0                         NaN               NaN   
175              24.5                         NaN               NaN   
202              40.0                         NaN               Na

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

> Заполнял сразу по всему набору данных

#### ПРИМЕЧАНИЕ
Домашнее задание сдается ссылкой на репозиторий [GitHub](https://github.com/).
Не сможем проверить или помочь, если вы пришлете:
- файлы;
- архивы;
- скриншоты кода.

Все обсуждения и консультации по выполнению домашнего задания ведутся только на соответствующем канале в slack.

##### Как правильно задавать вопросы аспирантам, преподавателям и коллегам
Прежде чем задать вопрос, попробуйте найти ответ в интернете. Навык самостоятельного поиска информации — один из важнейших. Каждый практикующий специалист любого уровня делает это ежедневно.

Сформулируйте вопрос по алгоритму:  
1) Что я делаю?  
2) Какого результата я ожидаю?  
3) Как фактический результат отличается от ожидаемого?  
4) Что я уже попробовал сделать, чтобы исправить проблему?  

По возможности прикрепите к вопросу скриншоты либо ссылки на код. Не выкладывайте все решение, оставляйте только проблемный и воспроизводимый участок кода.